src/Lykke.AlgoStore.MatchingEngineAdapter.Services/Listening/ListeningService.cs

```diff
@@ -0,0 +1,162 @@
+﻿using Lykke.AlgoStore.MatchingEngineAdapter.Core.Services;
+using System;
+using System.Net;
+using System.Net.Sockets;
+using Common.Log;
+using JetBrains.Annotations;
+using System.Threading.Tasks;
+using Lykke.AlgoStore.MatchingEngineAdapter.Abstractions.Services.Listening;
+using System.Collections.Concurrent;
+using Lykke.AlgoStore.CSharp.AlgoTemplate.Models.Repositories;
+using Lykke.AlgoStore.MatchingEngineAdapter.Core.Services.Listening;
+using System.Threading;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Lykke.AlgoStore.MatchingEngineAdapter.Services.Listening
+{
+    /// <summary>
+    /// Listens for incoming connections and passes them on to the load balancer
+    /// </summary>
+    public class ListeningService : IListeningService
+    {
+        // This will hold all of the connected instances.
+        // We use Dictionary here because there's no concurrent hashset in .NET. The value is not used
+        // and will be set to 0 for every connection
+        private readonly ConcurrentDictionary<string, byte> _connectionHashSet = new ConcurrentDictionary<string, byte>();
+        private readonly ConcurrentDictionary<Task, byte> _allWorkers = new ConcurrentDictionary<Task, byte>();
```

Comment:
Consider using https://github.com/LykkeCity/CommonDotNetLibraries/blob/master/src/Lykke.Common/TaskManager.cs here

Reply:
Not applicable, the task manager makes use of completion sources, while we need to use the tasks directly